### PR TITLE
Feat/wrpc

### PR DESCRIPTION
**Pull request by javierguerragiraldez**
wRPC protocol in CP/DP communications

- [x] connect via WebSockets, preserve existing authentication schemes.
- [x] bidirectional RPC, passing data/args/results/errors in both directions.
- [x] some error handling
- request/response/pending expiration
  - [x] drop expired packets on reception
  - [x] drop expired waiters on response
  - [ ] drop expired waiters that never receive response
- [x] encapsulate receive thread within protocol management
- [x] encapsulate send thread within protocol management, only for "server" sockets.
- [x] send periodic PINGs
- [ ] detect and act on missing pings or pongs.
- [x] send config data from CP to every DP on any change.
- [x] handle hash or version of config data, to ignore unneeded config messages.
- transform config data to fit protobuf spec
  - [x] wrap plugins config in Struct
  - [x] allow foreign keys to be passed in expanded form
  - [x] ensure "absent" fields don't get materialized with "default" or "empty" values on decoding.
  - ...
- [x] define inter-worker communication channel
- [x] permit calls from any worker
- [x] pass existing hybrid mode tests

Note: there's still some code duplication between the existing (json) protocol and this.  This is to ensure the feature flag (cluster_protocol=json|wrpc) doesn't activate any new code.  Once the version negotiation feature (#8397 ) is merged in, this will be refactored on top of that and repeated code will be factored away.

### CODE REVIEW

File: kong/tools/channel.lua [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+local RESP_VAL_KEY_PREFIX = "channel_resp_value_"
+
+
+local function waitstep(step, deadline)
+  sleep(step)
+  return min(max(0.001, step * 2), deadline-now(), 0.5)
+end
+
+
+local function add_wait(dict, key, val, exptime, deadline)
```
Comment (mikefero) [inline kong/tools/channel.lua:21]: WDYT about naming these `*_wait` functions to give a better description of what they are doing as it was not immediately obvious until looking through the function?

```suggestion
local function add_key_value_wait(dict, key, val, exptime, deadline)
```

I am not even a fan of this name `add_key_value_wait`; naming is hard.

File: kong/tools/wrpc.lua [hunk 2]
```diff
@@ -0,0 +1,748 @@
...
+    payloads = payloads,
+  })
+  return self.seq
+end
+
+--- little helper to ease grabbing an unspecified number
+--- of values after an `ok` flag
+local function ok_wrapper(ok, ...)
+  return ok, {n = select('#', ...), ...}
+end
```
Comment (mikefero) [inline kong/tools/wrpc.lua:435]: :open_mouth: This is magic!!!

File: kong/tools/wrpc.lua [hunk 3]
```diff
@@ -0,0 +1,748 @@
...
+wrpc_service.__index = wrpc_service
+
+
+--- a `service` object holds a set of methods defined
+--- in .proto files
+function wrpc.new_service()
+  if not wrpc_proto then
+    local wrpc_protofname = assert(proto_searchpath("wrpc.wrpc"))
+    wrpc_proto = assert(grpc.each_method(wrpc_protofname))
+    --pp("wrpc_proto", wrpc_proto)
```
Comment (mikefero) [inline kong/tools/wrpc.lua:219]: should this become a DEBUG log message?

File: kong/tools/wrpc.lua [hunk 4]
```diff
@@ -0,0 +1,748 @@
...
+
+
+--- Part of wrpc_peer:call()
+--- If calling the same method with the same args several times,
+--- (to the same or different peers), this method returns the
+--- invariant part, so it can be cached to reduce encoding overhead
+function wrpc_service:encode_args(name, ...)
+  local rpc = self:get_method(name)
+  if not rpc then
+    return nil, string.format("no method %q", name)
```
Comment (mikefero) [inline kong/tools/wrpc.lua:324]:
```suggestion
    return nil, string.format("unknown method %q", name)
```

**Review by mikefero — COMMENTED** [review]
Looks great so far; some nits and suggestions for now; feel free to punt some of them during another iteration.

First pass; need to look over `kong/tools/wrpc.lua` more though.

File: kong/clustering/init.lua [hunk 5]
```diff
@@ -127,7 +127,14 @@ function _M.new(conf)
...
 
-  self.child = require("kong.clustering." .. conf.role).new(self)
+  print("role: ", conf.role, "  protocol: ", conf.cluster_protocol)
+
+  local clustering_submodule = conf.role
+  if conf.cluster_protocol == "wRPC" then
+    clustering_submodule = "wrpc_" .. clustering_submodule
+  end
+
+  self.child = require("kong.clustering." .. clustering_submodule).new(self)
```
[the remark this answers is not in the document]
Comment (javierguerragiraldez) [reply kong/clustering/init.lua:137]: not a fan of using input data for filesystem search of code. it's after passing through the "enum" check, but still...

File: kong/clustering/wrpc_control_plane.lua [hunk 6]
```diff
@@ -0,0 +1,690 @@
...
+local CLUSTERING_SYNC_STATUS = constants.CLUSTERING_SYNC_STATUS
+local MAJOR_MINOR_PATTERN = "^(%d+)%.(%d+)%.%d+"
+local _log_prefix = "[clustering] "
+
+local wrpc_config_service
+
+local function get_config_service(self)
+  if not wrpc_config_service then
+    wrpc_config_service = wrpc.new_service()
+    wrpc_config_service:add("kong.services.config.v1.config")
```
[the remark this answers is not in the document]
Comment (javierguerragiraldez) [reply kong/clustering/wrpc_control_plane.lua:56]: a new protocol version would require new code.  i think it's best that each version of the code hardcodes the appropriate version of the .proto; in fact, .proto files are part of the code, even if loaded dynamically instead of generating static code as on Go

File: kong/clustering/wrpc_control_plane.lua [hunk 7]
```diff
@@ -0,0 +1,690 @@
...
+        minor   = minor,
+        version = version,
+      }
+
+    else
+      versions[name] = {}
+    end
+  end
+  return versions
+end
```
[the remark this answers is not in the document]
Comment (javierguerragiraldez) [reply kong/clustering/wrpc_control_plane.lua:116]: both `wrpc_control_plane.lua` and `wrpc_data_plane.lua` started as writing on top of `control_plane.lua` and `data_plane.lua` respectively.  when I got it mostly working, I renamed both and pulled the original ones back unchanged.

The idea is to make it **very** clear that the default case is the same, unchanged code as before.  After that gets accepted, we can refactor and even remove the flag in favor of version negotiation.

File: kong/clustering/wrpc_control_plane.lua [hunk 8]
```diff
@@ -0,0 +1,690 @@
...
+      self.plugins_configured[plugin.name] = true
+    end
+  end
+
+  config_version = config_version + 1
+
+  -- store serialized plugins map for troubleshooting purposes
+  local shm_key_name = "clustering:cp_plugins_configured:worker_" .. ngx.worker.id()
+  kong_dict:set(shm_key_name, cjson_encode(self.plugins_configured));
+  ngx_log(ngx_DEBUG, "plugin configuration map key: " .. shm_key_name .. " configuration: ", kong_dict:get(shm_key_name))
```
[the remark this answers is not in the document]
Comment (javierguerragiraldez) [reply kong/clustering/wrpc_control_plane.lua:166]: i don't think this is useful for debug logging, i put it when trying to understand how all this work.

File: kong/clustering/wrpc_control_plane.lua [hunk 9]
```diff
@@ -0,0 +1,690 @@
...
+    client.peer:send_encoded_call(self.config_call_rpc, self.config_call_args)
+
+    n = n + 1
+  end
+
+  ngx_log(ngx_NOTICE, _log_prefix, "config version #", config_version, " pushed to ", n, " clients")
+end
+
+
+function _M:validate_shared_cert()
```
[the remark this answers is not in the document]
Comment (javierguerragiraldez) [reply kong/clustering/wrpc_control_plane.lua:215]: yes, they're really not protocol-related.  i'm still a bit adverse on changing existing code this close to feature freeze

File: kong/clustering/wrpc_control_plane.lua [hunk 10]
```diff
@@ -0,0 +1,690 @@
...
+  end
+
+  -- TODO: DAOs are not checked in any way at the moment. For example if plugin introduces a new DAO in
+  --       minor release and it has entities, that will most likely fail on data plane side, but is not
+  --       checked here.
+
+  return true, nil, CLUSTERING_SYNC_STATUS.NORMAL
+end
+
+function _M:handle_cp_websocket()
```
[the remark this answers is not in the document]
Comment (javierguerragiraldez) [reply kong/clustering/wrpc_control_plane.lua:418]: this function is exported and called from `clustering/init.lua` for both protocols, makes sense to keep the same name/signature

File: kong/tools/channel.lua (hunk 1 above)
Comment (javierguerragiraldez) [reply kong/tools/channel.lua:21]: yeah, naming is hard.  basically these are the same as the shdict `:add()` and `:get()` methods but which block the coroutine until the operation gets done (or timeout).  i'll add some doccoments there.

File: kong/tools/wrpc.lua (hunk 2 above)
Comment (javierguerragiraldez) [reply kong/tools/wrpc.lua:435]: i don't like it too much, it's there to preserve the plurality of the `payloads` argument, both as input and output

File: kong/tools/wrpc.lua (hunk 3 above)
Comment (javierguerragiraldez) [reply kong/tools/wrpc.lua:219]: i though i had removed all these!

File: kong/clustering/wrpc_control_plane.lua [hunk 11]
```diff
@@ -0,0 +1,690 @@
...
+      end
+
+    elseif err ~= "timeout" then
+      ngx_log(ngx_ERR, _log_prefix, "semaphore wait error: ", err)
+    end
+  end
+end
+
+
+function _M:init_worker()
```
[the remark this answers is not in the document]
Comment (javierguerragiraldez) [reply kong/clustering/wrpc_control_plane.lua:614]: in this case i think further evolution would make these more different than right now.  maybe as soon as with the addition of version negotiation.

**Review by mikefero — APPROVED** [review]
Great work! Let's get this in and then iterate on the remaining items in your list and then we can focus on other issues as them come in. With this code being completely isolated from the JSON payload we will be in good shape.

File: kong/clustering/wrpc_control_plane.lua [hunk 12]
```diff
@@ -0,0 +1,690 @@
...
+local function remove_nulls(tbl)
+  for k,v in pairs(tbl) do
+    if v == ngx_null then
+      tbl[k] = nil
+    elseif type(v) == "table" then
+      tbl[k] = remove_nulls(v)
+    end
+  end
+  return tbl
+end
```
[the remark this answers is not in the document]
Comment (javierguerragiraldez) [reply kong/clustering/wrpc_control_plane.lua:143]: i haven't checked too much into `kong.db.declarative`, as I want it to be fully compatible with the existing data flow.... for now.  I'd much prefer that `:export_config()` didn't add those `json.null` or `ngx.null` objects.

File: kong/conf_loader/init.lua [hunk 13]
```diff
@@ -636,6 +636,7 @@ local CONF_INFERENCES = {
   lua_socket_pool_size = { typ = "number" },
 
   role = { enum = { "data_plane", "control_plane", "traditional", }, },
+  cluster_protocol = { enum = { "json", "wRPC" }, },
```
[the remark this answers is not in the document]
Comment (javierguerragiraldez) [reply kong/conf_loader/init.lua]: done

**Review by mikefero — APPROVED** [review]

File: kong/include/kong/services/legal/README.md [hunk 14]
```diff
@@ -0,0 +1,16 @@
+# Legal service
+
```
Comment (hbagdi) [inline kong/include/kong/services/legal/README.md:2]: Please remove this. This was not supposed to be part of OSS.

File: kong/include/kong/services/metrics/v1/README.md [hunk 15]
```diff
@@ -0,0 +1,3 @@
+## Metrics
+
+Metrics service is used by a DP to report back its metrics to the CP.
```
Comment (hbagdi) [inline kong/include/kong/services/metrics/v1/README.md:3]: Please remove this service as well, this was not supposed to be part of OSS.

File: kong/include/kong/services/event_hooks/v1/event_hooks.proto [hunk 16]
```diff
@@ -0,0 +1,43 @@
+syntax = "proto3";
+
```
Comment (hbagdi) [inline kong/include/kong/services/event_hooks/v1/event_hooks.proto:2]: This file was never supposed to be part of OSS. Please remove it.

File: kong/include/kong/model/events/event_hooks.proto [hunk 17]
```diff
@@ -0,0 +1,18 @@
+syntax = "proto3";
+
```
Comment (hbagdi) [inline kong/include/kong/model/events/event_hooks.proto:2]: Please remove this.

File: kong/include/kong/services/legal/README.md (hunk 14 above)
Comment (javierguerragiraldez) [reply kong/include/kong/services/legal/README.md]: done

File: kong/include/kong/services/event_hooks/v1/event_hooks.proto (hunk 16 above)
Comment (javierguerragiraldez) [reply kong/include/kong/services/event_hooks/v1/event_hooks.proto]: done

File: kong/include/kong/services/metrics/v1/README.md (hunk 15 above)
Comment (javierguerragiraldez) [reply kong/include/kong/services/metrics/v1/README.md]: done

File: kong/include/kong/model/events/event_hooks.proto (hunk 17 above)
Comment (javierguerragiraldez) [reply kong/include/kong/model/events/event_hooks.proto]: done

File: kong/include/kong/model/target.proto [hunk 18]
```diff
@@ -0,0 +1,16 @@
+syntax = "proto3";
+
+option go_package = "github.com/kong/koko/internal/gen/wrpc/kong/model;model";
+
+package kong.model;
+
+import "kong/model/upstream.proto";
+
+message Target {
+  double created_at = 1;
```
Comment (zhongchen) [inline kong/include/kong/model/target.proto:10]: the `type` should be `int32` instead of `double`? it is defined as `int32` in all other places.

Comment (javierguerragiraldez) [reply kong/include/kong/model/target.proto:10]: i've forwarded the fix to the original repo (these files are copied there, ideally should be copied during build time and not on the repo.  we'll get there soon)

File: kong/tools/wrpc.lua [hunk 19]
```diff
@@ -0,0 +1,686 @@
...
+local wrpc_service = {}
+wrpc_service.__index = wrpc_service
+
+
+--- a `service` object holds a set of methods defined
+--- in .proto files
+function wrpc.new_service()
+  if not wrpc_proto then
+    local wrpc_protofname = assert(proto_searchpath("wrpc.wrpc"))
+    wrpc_proto = assert(grpc.each_method(wrpc_protofname))
```
Comment (Suika-Kong) [inline kong/tools/wrpc.lua:135]: It is a bit confusing that wRPC should call gRPC's method. Maybe we can extract the protobuf part out?

File: kong/clustering/wrpc_control_plane.lua [hunk 20]
```diff
@@ -0,0 +1,662 @@
...
+    _, err, client.sync_status = self:check_version_compatibility(dp_version, client.dp_plugins_map, log_suffix)
+    if err then
+      ngx_log(ngx_ERR, _log_prefix, err, log_suffix)
+      wb:send_close()
+      client:update_sync_status()
+      return ngx_exit(ngx_CLOSE)
+    end
+  end
+
+  self:push_config()    -- first config push
```
Comment (Suika-Kong) [inline kong/clustering/wrpc_control_plane.lua:528]: Won't this cause a config push to all nodes for every new connection?
Maybe this would do.

```suggestion
  do
    local payload, err = self:export_deflated_reconfigure_payload()
    if not payload then
      ngx_log(ngx_ERR, _log_prefix, "unable to export config from database: ", err)
      return
    end
    ngx_log(ngx_NOTICE, _log_prefix, "config version #", config_version, " pushed to ", dp_id)
  end
```

Comment (Suika-Kong) [reply kong/clustering/wrpc_control_plane.lua:528]: Also, maybe we can cache the "payload" to speed up the case, where new nodes come and the config is not modified.

File: kong/clustering/wrpc_control_plane.lua [hunk 21]
```diff
@@ -0,0 +1,662 @@
...
+  end
+
+  local n = 0
+  for _, client in pairs(self.clients) do
+    client.peer:send_encoded_call(self.config_call_rpc, self.config_call_args)
+
+    n = n + 1
+  end
+
+  ngx_log(ngx_NOTICE, _log_prefix, "config version #", config_version, " pushed to ", n, " clients")
```
Comment (Suika-Kong) [inline kong/clustering/wrpc_control_plane.lua:183]: I notice that in the non-wrpc version, we log in a different format. Should we unify them?

Comment (Suika-Kong) [reply kong/clustering/wrpc_control_plane.lua:183]: Also, the log level is different.

File: kong/clustering/wrpc_data_plane.lua [hunk 22]
```diff
@@ -0,0 +1,374 @@
...
+  end
+
+  local entities, err, _, meta, new_hash =
+              self.declarative_config:parse_table(config_table, config_hash)
+  if not entities then
+    return nil, "bad config received from control plane " .. err
+  end
+
+  if declarative.get_current_hash() == new_hash then
+    ngx_log(ngx_INFO, _log_prefix, "same config received from control plane, ",
```
Comment (Suika-Kong) [inline kong/clustering/wrpc_data_plane.lua:82]: Ditto.

File: kong/tools/grpc.lua [hunk 23]
```diff
@@ -13,7 +13,7 @@ local grpc = {}
 
 local function safe_set_type_hook(type, dec, enc)
   if not pcall(pb.hook, type) then
-    ngx.log(ngx.NOTICE, "no type '" .. type .. "' defined")
+    ngx.log(ngx.DEBUG, "no type '" .. type .. "' defined")
```
Comment (Suika-Kong) [inline kong/tools/grpc.lua:16]: Log level changed.

**Review by Suika-Kong — COMMENTED** [review]
And maybe more refactoring(extracting common codes) also helps when implementing the version negotiation.

File: kong/clustering/wrpc_data_plane.lua [hunk 24]
```diff
@@ -0,0 +1,374 @@
...
+        " (retrying after ", reconnection_delay, " seconds)", log_suffix)
+
+      assert(ngx.timer.at(reconnection_delay, function(premature)
+        self:communicate(premature)
+      end))
+      return
+    end
+  end
+
+  local config_semaphore = semaphore.new(0)
```
[the remark this answers is not in the document]
Comment (javierguerragiraldez) [reply kong/clustering/wrpc_data_plane.lua:256]: good idea.  the whole `peer` object should be dropped if the connection is lost, but not sure if semaphore objects are "sticky" and might avoid GC.

File: kong/clustering/wrpc_control_plane.lua (hunk 20 above)
Comment (javierguerragiraldez) [reply kong/clustering/wrpc_control_plane.lua:528]: good call.

File: kong/tools/grpc.lua (hunk 23 above)
Comment (javierguerragiraldez) [reply kong/tools/grpc.lua:16]: yes, this message appears constantly because the protobuf state is global; but in most cases can be safely ignored so i wanted to quiet it down a bit.  still don't want to remove it because it _is_ useful when adding new "special types"

File: kong/clustering/wrpc_data_plane.lua (hunk 24 above)
[the remark this answers is not in the document]
Comment (javierguerragiraldez) [reply kong/clustering/wrpc_data_plane.lua:256]: ... on the other hand there's no `semaphore.reset()`.  i'll just add a `peer.semaphore = nil` to be sure it's removed as soon as possible.

File: kong/clustering/wrpc_control_plane.lua (hunk 20 above)
Comment (javierguerragiraldez) [reply kong/clustering/wrpc_control_plane.lua]: done!

**Review by Suika-Kong — APPROVED** [review]

**Comment by javierguerragiraldez** [comment]
rebased on top of master

File: kong/clustering/wrpc_data_plane.lua [hunk 25]
```diff
@@ -0,0 +1,379 @@
...
+  local peer = wrpc.new_peer(c, get_config_service(), { channel = true })
```
Comment (zhongchen) [inline kong/clustering/wrpc_data_plane.lua:258]: What's the best way for me to share a wrpc connection among Ngnix workers? Should each worker create its own peer or should I share the peer among workers?

Comment (zhongchen) [reply kong/clustering/wrpc_data_plane.lua:258]: I need to let each Nginx worker push its local data to the cp.

File: kong/clustering/wrpc_data_plane.lua [hunk 26]
```diff
@@ -0,0 +1,379 @@
...
+  end
+
+  -- Here we spawn two threads:
+  --
+  -- * config_thread: it grabs a received declarative config and apply it
+  --                  locally. In addition, this thread also persists the
+  --                  config onto the local file system
+  -- * ping_thread: performs a ConfigService.PingCP call periodically.
+
+  local ping_immediately
```
Comment (zhongchen) [inline kong/clustering/wrpc_data_plane.lua:289]: `ping_immediately` is never set in this file, while this field is set in `data_plane.lua`.

File: kong/clustering/wrpc_data_plane.lua (hunk 25 above)
Comment (javierguerragiraldez) [reply kong/clustering/wrpc_data_plane.lua:258]: one worker owns the connection and creates a peer.  the other workers create a remote_peer, which use a channel to communicate with the "real" one.

right now the channel name is a hardcoded constant, so there can only be one "main" peer (the one created in wrpc_data_plane) and all remote_peers go to that single one.  so anywhere in the code you can use `wrpc.new_remote_peer()` and it will use it.   if we require separate wrpc connections, we'll have to pass the channel name somewhere.

note, the "main" peer doesn't encode/decode the remote_peer's data, so you can add new service definitions and don't have to add to the "main" peer, just the remote_peer

File: kong/clustering/wrpc_data_plane.lua (hunk 26 above)
Comment (javierguerragiraldez) [reply kong/clustering/wrpc_data_plane.lua:289]: good catch.  in the original `data_plane.lua` it's used to send the first ping with the config hash and prime the config pushing from the control_plane; but in wrpc the config is initiated by the control plane and this isn't required.

i'll remove the variable

File: kong/clustering/wrpc_data_plane.lua (hunk 25 above)
Comment (zhongchen) [reply kong/clustering/wrpc_data_plane.lua:258]: I will create another wrpc connection for the telemetry endpoint.

Comment (zhongchen) [reply kong/clustering/wrpc_data_plane.lua:258]: There will only be one wrpc connection for each data plane seems to be a very restrictive assumption.

Comment (javierguerragiraldez) [reply kong/clustering/wrpc_data_plane.lua:258]: then i'll have to add a "connection name" and move the current constant from `wrcp.lua` to `wrpc_data_plane.lua`

Comment (zhongchen) [reply kong/clustering/wrpc_data_plane.lua:258]: I am wondering why do you need a channel? It seems that you never create a `remote_client`. no usage of `wprc.new_remote_client`

Comment (javierguerragiraldez) [reply kong/clustering/wrpc_data_plane.lua:258]: the config handling is done completely in a single worker so it doesn't need it; but the wrpc has to be usable from any worker, so i added channel.lua and new_remote_client()

Comment (zhongchen) [reply kong/clustering/wrpc_data_plane.lua:258]: Cool. I was just curious why you enabled the `channel` for the config connection, while it is not needed. https://github.com/Kong/kong/blob/7e9a9a28e82a7a58a3c501f05f241b137f676f69/kong/clustering/wrpc_data_plane.lua#L257

Comment (javierguerragiraldez) [reply kong/clustering/wrpc_data_plane.lua:258]: it's expected that any code would have to initiate calls to the CP.  this allows any part of code to just do it without having to deal with connection handling.

Comment (javierguerragiraldez) [reply kong/clustering/wrpc_data_plane.lua]: done

File: kong/clustering/wrpc_data_plane.lua [hunk 27]
```diff
@@ -0,0 +1,374 @@
...
+      self:communicate(premature)
+    end))
+  end
+end
+
+
+local wrpc_config_service
+local function get_config_service()
+  if not wrpc_config_service then
+    wrpc_config_service = wrpc.new_service()
```
Comment (zhongchen) [inline kong/clustering/wrpc_data_plane.lua:182]: do all ngnix workers need to share the same service or each worker can create its own?

Comment (zhongchen) [reply kong/clustering/wrpc_data_plane.lua:182]: I need to do something like
```
    self.peer = wrpc.new_remote_client(get_requests_log_service(), self.TELEMETRY_CHANNEL_NAME)
```

Comment (zhongchen) [reply kong/clustering/wrpc_data_plane.lua:182]: My understanding is that `local` means each worker will create its own `wrpc_requests_log_service`

Comment (javierguerragiraldez) [reply kong/clustering/wrpc_data_plane.lua:182]: > ... share the same service ...
the service is a Lua object.  as such, it can't be shared between workers.

yes, each worker will need its own, like everything in openresty.

File: kong/clustering/wrpc_data_plane.lua [hunk 28]
```diff
@@ -0,0 +1,374 @@
...
+local inflate_gzip = utils.inflate_gzip
+local deflate_gzip = utils.deflate_gzip
+
+
+local KONG_VERSION = kong.version
+local CONFIG_CACHE = ngx.config.prefix() .. "/config.cache.json.gz"
+local ngx_ERR = ngx.ERR
+local ngx_DEBUG = ngx.DEBUG
+local ngx_INFO = ngx.INFO
+local MAX_PAYLOAD = constants.CLUSTERING_MAX_PAYLOAD
```
Comment (zhongchen) [inline kong/clustering/wrpc_data_plane.lua:35]: `CLUSTERING_MAX_PAYLOAD` is not part of the constants anymore. https://github.com/Kong/kong/pull/8337/files

File: kong/tools/wrpc.lua [hunk 29]
```diff
@@ -0,0 +1,685 @@
...
+    if typ == "close" then
+      kong.log.notice("Received WebSocket \"close\" frame from peer: ", err, ": ", data)
+      return self:close()
+    end
+  end
+end
+
+--- RPC call.
+--- returns the call sequence number, doesn't wait for response.
+function wrpc_peer:call(name, ...)
```
[the remark this answers is not in the document]
Comment (javierguerragiraldez) [reply kong/tools/wrpc.lua:315]: i guess you're asking about remote_peers.  they have their own implementation of `:call()`

[the remark this answers is not in the document]
Comment (zhongchen) [reply kong/tools/wrpc.lua:315]: so I should do something like
```
assert(self.peer:remote_call("RequestsLogService.SendLogs", { request_metadata = request_metadata }))
``` for remote peers
and
```
assert(self.peer:call("RequestsLogService.SendLogs", { request_metadata = request_metadata }))
``` for the peer?

**Comment by Suika-Kong** [comment]
Your PR seems ready to merge, but there are some lint issues that fail checks. May you spend some time fixing that?

File: kong/templates/kong_defaults.lua [hunk 30]
```diff
@@ -22,6 +22,7 @@ admin_listen = 127.0.0.1:8001 reuseport backlog=16384, 127.0.0.1:8444 http2 ssl
 status_listen = off
 cluster_listen = 0.0.0.0:8005
 cluster_control_plane = 127.0.0.1:8005
+cluster_protocol = json
```
[the remark this answers is not in the document]
Comment (hbagdi) [reply kong/templates/kong_defaults.lua:24]: Is this the right default?

File: kong/tools/wrpc.lua [hunk 31]
```diff
@@ -0,0 +1,685 @@
...
+
+        elseif keyword == "rpc" then
+          identifier = scope_name .. "." .. identifier
+        end
+
+        local type_annotations = annotations[keyword]
+        if type_annotations then
+          local tag_key, tag_value = annotation:match("^%s*(%S-)=(%S+)%s*$")
+          if tag_key and tag_value then
+            tag_value = tag_value
```
Comment (Suika-Kong) [inline kong/tools/wrpc.lua:172]: It seems of no effect?

Comment (javierguerragiraldez) [reply kong/tools/wrpc.lua:172]: right.  seems a leftover from debugging code

File: kong/tools/channel.lua [hunk 32]
```diff
@@ -0,0 +1,164 @@
...
+
+  return nil, err
+end
+
+
+--- Waits until a value is posted by any client
+--- @param dict shdict shdict to use
+--- @return any, string, string value, channel name, error
+function Channel.wait_all(dict)
+  local name, err = get_wait(dict, NAME_KEY, now() + DEFAULT_TIMEOUT)
```
Comment (Suika-Kong) [inline kong/tools/channel.lua:138]: Could we just use one dictionary key, and put the name of the channel together into the value? So we could fetch only once.
After all, post keeps blocked if `NAME_KEY` is not free.

File: kong/tools/wrpc.lua [hunk 33]
```diff
@@ -0,0 +1,685 @@
...
+    end
+  end
+  proto_f:close()
+
+  grpc.each_method(service_fname, function(_, srvc, mthd)
+    assert(srvc.name)
+    assert(mthd.name)
+    local rpc_name = srvc.name .. "." .. mthd.name
+
+    local service_id = assert(annotations.service[srvc.name] and annotations.service[srvc.name]["service-id"])
```
Comment (Suika-Kong) [inline kong/tools/wrpc.lua:188]: Could we just do this?
```lua
If service_id[srvc.name] == nil then
  service_id[srvc.name] =  new_service_id
  new_service_id = new_service_id +1
end
```

File: kong/tools/wrpc.lua [hunk 34]
```diff
@@ -0,0 +1,685 @@
...
+        end
+
+        local type_annotations = annotations[keyword]
+        if type_annotations then
+          local tag_key, tag_value = annotation:match("^%s*(%S-)=(%S+)%s*$")
+          if tag_key and tag_value then
+            tag_value = tag_value
+            local tags = type_annotations[identifier] or {}
+            type_annotations[identifier] = tags
+            tags[tag_key] = tag_value
```
Comment (Suika-Kong) [inline kong/tools/wrpc.lua:175]: tags seem not used

File: kong/tools/channel.lua (hunk 32 above)
Comment (javierguerragiraldez) [reply kong/tools/channel.lua:138]: this is one of the reasons to move to a socket implementation, it's hard to follow which concepts tie 1:1 into which and which are 1:n
in this case, i think the purpose was to reduce serialization of processes, each client pushes their value without blocking each other, and only block on making the "server" pull the notification.    at higher level i think there's still the same contention anyway... but it's important to be able for each client to fire the message and not wait for all the others to be served.

File: kong/tools/wrpc.lua (hunk 33 above)
Comment (javierguerragiraldez) [reply kong/tools/wrpc.lua:188]: you mean dynamically assign ids to methods and rpcs?  no, we must respect those in the `.proto` file

File: kong/tools/wrpc.lua (hunk 34 above)
Comment (javierguerragiraldez) [reply kong/tools/wrpc.lua:175]: they're part of the `annotation` structure, it's put into `type_annotations`, which is a subtable of `annotations`.  after scanning the file, it's read to get the `service_id` and `rpc_id` on the `.each_method()` loop

File: kong/tools/wrpc.lua [hunk 35]
```diff
@@ -0,0 +1,685 @@
...
+  rpc.response_handler = response_handler
+  return rpc
+end
+
+
+--- Part of wrpc_peer:call()
+--- If calling the same method with the same args several times,
+--- (to the same or different peers), this method returns the
+--- invariant part, so it can be cached to reduce encoding overhead
+function wrpc_service:encode_args(name, ...)
```
Comment (Suika-Kong) [inline kong/tools/wrpc.lua:236]: It seems that all call to `call` and `call_wait` uses a table to contain arguments.

File: kong/tools/wrpc.lua [hunk 36]
```diff
@@ -0,0 +1,685 @@
...
+  local rpc = self:get_method(name)
+  if not rpc then
+    return nil, string.format("unknown method %q", name)
+  end
+
+  local num_args = select('#', ...)
+  local payloads = table.new(num_args, 0)
+  for i = 1, num_args do
+    payloads[i] = assert(pb.encode(rpc.input_type, select(i, ...)))
```
Comment (Suika-Kong) [inline kong/tools/wrpc.lua:245]: Encode every arguments with the same `input_type` seems a bug.

File: kong/tools/wrpc.lua (hunk 35 above)
Comment (javierguerragiraldez) [reply kong/tools/wrpc.lua:236]: not sure what do you mean.   the specific case i had in mind is the CP sending the same config to several DP nodes.  instead of repeating the call with the unencoded config, it calls `service:encode_args()` once, and then `peer:send_encoded_call()` for each DP node

File: kong/tools/wrpc.lua (hunk 36 above)
Comment (javierguerragiraldez) [reply kong/tools/wrpc.lua:245]: the RPC definitions in .proto files only specify one input type.   still, the frame structure supports repeated payloads.  a single call could specify several items of the same time instead of using an array.

File: kong/tools/wrpc.lua (hunk 35 above)
Comment (Suika-Kong) [reply kong/tools/wrpc.lua:236]: I mean like this:
```lua
local resp, err = peer:call_wait("ConfigService.ReportMetadata", { plugins = self.plugins_list })
assert(peer:call("ConfigService.PingCP", { hash = hash }))
```
This comment is related to the comment below.

File: kong/tools/wrpc.lua (hunk 36 above)
Comment (Suika-Kong) [reply kong/tools/wrpc.lua:245]: Maybe I misinterpreted the comment in the wrpc.proto:
```protobuf
  // payloads is an array representing the request or response data of an RPC.
  // A request message MAY contain multiple elements where each element represents
  // an argument to the RPC call. The order of the elements MUST correspond to
  // the order of arguments in the function call.
  // A response message MUST contain a single payload.
  // Use a wrapper type for RPCs which contain multiple responses.
  // Unless otherwise specified by the Service or RPC, the encoding method in
  // use is PROTO3.
  //
  // Note: This results in double proto3 encoding. Once encoding of payloads,
  // and then encoding of the entire message. We acknowledge that there are
  // some overheads here and they will be addressed in a future iteration.
  repeated bytes payloads = 9;
```

By "each element represents an argument" I thought it was for every member of the request message (like "hash" and "plugins" in the previous comment).
Also, if an RPC call receives repeated messages, it cannot use repeat like this:

```proto
  rpc SomeCall(repeated SomeCallRequest) returns (SomeCallResponse);
```

But to use something like:

```proto
Message RepeatedSomeCallRequest {
    repeated SomeCallRequest arguments = 1;
}
```

So I don't think it makes sense to encode an array of payloads...

Comment (javierguerragiraldez) [reply kong/tools/wrpc.lua:245]: > Also, if an RPC call receives repeated messages, it cannot use repeat like this:
>
>  rpc SomeCall(repeated SomeCallRequest) returns (SomeCallResponse);

right, the syntax doesn't support this.  but in the current implementation, the "repeated" is implied, so this could be done:

```protobuf
rpc SomeCall(SomeCallRequest) returns (someCallResponse);

message SomeCallRequest {
    string name = 1;
    string value = 2;
}
```
and in Lua:
```lua
   peer:call("serv.SomeCall", {name="one", value=1}, {name="two", value=2}, {name="three", value=3})
```

and it would result in a single call with three payload elements of the same type.

But I agree that it's not an important usecase, and the structure you describe would be just as good.

Comment (Suika-Kong) [reply kong/tools/wrpc.lua:245]: @hbagdi May you clarify this? Do we really want to implement it this way?
